File: 6.迭代器模式/geqiangbao_20201125_iterator_OC/geqiangbao_20201125_iterator_OC/demo/ArrayIterator.m
//
//  ArrayIterator.m
//  geqiangbao_20201125_iterator_OC
//
//  Created by xiaobao on 2020/11/25.
//

#import "ArrayIterator.h"

@implementation ArrayIterator

- (id)next{
    if ([self hasNext]) {
       id result = self.datas[self.index++];
        return result;
    }
    return nil;
   
}

- (BOOL)hasNext{
//    self.datas.count
//    return YES;
    return self.index < self.datas.count;
}



@end

File: 6.迭代器模式/geqiangbao_20201125_iterator_swift/geqiangbao_20201125_iterator_swift/demo/ArrayIterator.swift
//
//  ArrayIterator.swift
//  geqiangbao_20201125_iterator_swift
//
//  Created by xiaobao on 2020/11/25.
//

import UIKit

class ArrayIterator<M>: AnyIterator<Array<M>,M> {

    override func hasNext() -> Bool {
        return self.index < self.datas?.count ?? 0
    }
    
    override func next() -> M? {
        if self.hasNext() {
            let result = self.datas?[self.index]
            self.index += 1
            return result
        } else {
            return nil
        }
    }
}
